java/arcs/sdk/jvm/SingletonImpl.kt

```diff
@@ -20,16 +20,23 @@ class SingletonImpl<T : Entity>(
     entitySpec: EntitySpec<T>
 ) : ReadWriteSingleton<T> {
     private var entity: T? = null
+    private var onUpdateAction: (T?) -> Unit = {}
 
     override fun get(): T? = entity
 
     override fun set(entity: T) {
         this.entity = entity
         particle.onHandleUpdate(this)
+        onUpdateAction(entity)
     }
 
     override fun clear() {
         this.entity = null
         particle.onHandleUpdate(this)
+        onUpdateAction(entity)
+    }
+
+    override fun onUpdate(action: (T?) -> Unit) {
+        onUpdateAction = action
```

Comment:
I think this should be a list of callbacks, not a single one.

Reply:
Could you please at least add an assert of some sort that only one callback is being set?